### ISSUE TITLE
main

### ISSUE THREAD

**Issue by congmonkey** [bob48523/MultiTask-Learning#1]
一个参考的:https://discuss.pytorch.org/t/how-to-create-model-with-sharing-weight/398/2
# -*- coding: utf-8 -*-

# License: BSD
# Author: Sasank Chilamkurthy

from __future__ import print_function, division

import argparse

from models import *
import config as cf
from preprocess import *

import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd import Variable
import torch.backends.cudnn as cudnn

import torchvision
from torchvision import datasets, models, transforms

import matplotlib.pyplot as plt
import time
import copy
import os
import setproctitle
import numpy as np

os.environ["CUDA_VISIBLE_DEVICES"] = "5"

parser = argparse.ArgumentParser(description='PyTorch VGG Training')
parser.add_argument('--resume', '-r', action='store_true', help='resume from checkpoint')
parser.add_argument('--save')
parser.add_argument('--testOnly', '-t', action='store_true', help='Test mode with the saved model')

args = parser.parse_args()
args.save = args.save or './joint.base'
setproctitle.setproctitle(args.save)

use_cuda = torch.cuda.is_available()
best_acc = 0  # best test accuracy
start_epoch = cf.start_epoch  # start from epoch 0 or last checkpoint epoch
batch_size = cf.batch_size
epochs = cf.num_epochs
######################################################################
# Load Data

data_dir = {
    'cifar': '/ssd/RookieProject/CIFAR100/',
    'tiny': '/ssd/RookieProject/TinyImageNet',
    'vgg': '/ssd/RookieProject/VGGFlowers',
}

test_transforms = {
    x: transforms.Compose([
        transforms.RandomSizedCrop(64),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        #transforms.Normalize(cf.mean[x], cf.std[x])
    ])
    for x in ['cifar', 'tiny', 'vgg']
}

val_transforms = {
    x: transforms.Compose([
        transforms.Scale(72),
        transforms.CenterCrop(64),
        transforms.ToTensor(),
        #transforms.Normalize(cf.mean[x], cf.std[x])
    ])
    for x in ['cifar', 'tiny', 'vgg']
}

trainset = {
    'cifar': TxtFolder(os.path.join(data_dir['cifar']), 'list_train.txt', test_transforms['cifar']),
    'tiny': datasets.ImageFolder(os.path.join(data_dir['tiny'], 'train'), test_transforms['tiny']),
    'vgg': datasets.ImageFolder(os.path.join(data_dir['vgg'], 'train'), test_transforms['vgg']),
}

valset = {
    'cifar': TxtFolder(os.path.join(data_dir['cifar']), 'list_val.txt', val_transforms['cifar']),
    'tiny': datasets.ImageFolder(os.path.join(data_dir['tiny'], 'val'), val_transforms['tiny']),
    'vgg': datasets.ImageFolder(os.path.join(data_dir['vgg'], 'val'), val_transforms['vgg']),
}

trainloader = {
    x: torch.utils.data.DataLoader(trainset[x], batch_size=cf.batch_size,
                                   shuffle=True, num_workers=4)
    for x in ['cifar', 'tiny', 'vgg']
}

valloader = {
    x: torch.utils.data.DataLoader(valset[x], batch_size=cf.batch_size,
                                   shuffle=False, num_workers=4)
    for x in ['cifar', 'tiny', 'vgg']
}

use_gpu = torch.cuda.is_available()

class AverageMeter(object):
    """Computes and stores the average and current value"""
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count
######################################################################
# Training the model
if (args.testOnly):
    print('\n[Test Phase] : Model setup')
    #assert os.path.isdir('checkpoint'), 'Error: No checkpoint directory found!'
    checkpoint = torch.load(os.path.join(args.save, 'cifar_resnetse_best.t7'))
    net = checkpoint['net']
    
    if use_cuda:
        net.cuda()
       # net = torch.nn.DataParallel(net, device_ids=range(torch.cuda.device_count()))
       # cudnn.benchmark = True

    net.eval()

    for data_name in ['cifar','tiny','vgg']:
        test_loss = 0
        correct = 0
        total = 0        
        for batch_idx, (inputs, targets) in enumerate(valloader[data_name]):
            if use_cuda:
                inputs, targets = inputs.cuda(), targets.cuda()
            inputs, targets = Variable(inputs, volatile=True), Variable(targets)
            cifaro, tinyo, vggo = net(inputs, inputs, inputs)

            if data_name == 'cifar':
                outputs = cifaro
            elif data_name == 'tiny':
                outputs = tinyo
            elif data_name == 'vgg':
                outputs = vggo
                
            _, predicted = torch.max(outputs.data, 1)
            total += targets.size(0)
            correct += predicted.eq(targets.data).cpu().sum()

    # Save checkpoint when best model
        acc = 100.*correct/total
        print("| Test " + data_name +" Result\tAcc@1: %.2f%%" %(acc))

    sys.exit(0)
    
def train(net, criterion, epoch, trainF, data_name):
    print('\nEpoch: %d' % epoch)
    net.train()
    train_loss = 0
    top1 = AverageMeter()
    top5 = AverageMeter()
    losses = AverageMeter()
    k_cifar = 1
    k_tiny = 1
    k_vgg = 1
#    correct = 0
    total = 0
    nProcessed = 0
    
    nTrain = len(trainloader[data_name].dataset)

    optimizer = optim.SGD(net.parameters(), lr=cf.adjust_rate(epoch), momentum=cf.momentum, weight_decay=1e-4)
    
    for batch_idx, (inputs, targets) in enumerate(trainloader[data_name]):
        if use_cuda:
            inputs, targets = inputs.cuda(), targets.cuda()
        optimizer.zero_grad()
        target = targets
        inputs, targets = Variable(inputs), Variable(targets)
        
        cifaro, tinyo, vggo = net(inputs, inputs, inputs)
        if data_name == 'cifar':
            loss = criterion(cifaro, targets)
        elif data_name == 'tiny':
            loss = criterion(tinyo, targets)
        elif data_name == 'vgg':
            loss = criterion(vggo, targets) 
            
        #make_graph.save('./t.dot', loss.creator); assert(False)
        loss.backward()
        optimizer.step()
        
        if data_name == 'cifar':
            outputs = cifaro

        elif data_name == 'tiny':
            outputs = tinyo
            
        elif data_name == 'vgg':
            outputs = vggo
            
        nProcessed += len(inputs)
        train_loss += loss.data[0]
        _, predicted = torch.max(outputs.data, 1)
        total += targets.size(0)
#        correct += predicted.eq(targets.data).cpu().sum()
        prec1, prec5 = accuracy(outputs.data, target, topk=(1, 5))
        top1.update(prec1[0], inputs.size(0))
        top5.update(prec5[0], inputs.size(0))
        losses.update(loss.data[0], inputs.size(0))
        incorrect = predicted.ne(targets.data).cpu().sum()
        err = 100.*incorrect/len(inputs)
        partialEpoch = epoch + batch_idx / len(trainloader[data_name]) - 1
        
        print('Train Epoch: {:.2f} [{}/{} ({:.0f}%)]\tLoss: {:.6f}\tError: {:.6f}'.format(
            partialEpoch, nProcessed, nTrain, 100. * batch_idx / len(trainloader[data_name]),
            loss.data[0], err))
        print(' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}'
              .format(top1=top1, top5=top5))
        
    trainF.write('{},{},{},{}\n'.format(epoch, losses.avg, top1.avg, top5.avg))
    trainF.flush()
        
def test(net, epoch, testF, data_name):
    global best_acc
    net.eval()
    top1 = AverageMeter()
    top5 = AverageMeter()
    test_loss = 0
    incorrect = 0
    correct = 0
    total = 0 
    k = {
     'cifar': 0.33,
     'tiny': 0.33,
     'vgg': (1-0.66),
    }
    for batch_idx, (inputs, targets) in enumerate(valloader[data_name]):
        if use_cuda:
            inputs, targets = inputs.cuda(), targets.cuda()
        target = targets
        inputs, targets = Variable(inputs, volatile=True), Variable(targets)
        cifaro, tinyo, vggo = net(inputs, inputs, inputs)
            
        if data_name == 'cifar':
            loss = criterion(cifaro, targets)
        elif data_name == 'tiny':
            loss = criterion(tinyo, targets)
        elif data_name == 'vgg':
            loss = criterion(vggo, targets) 

        if data_name == 'cifar':
            outputs = cifaro

        elif data_name == 'tiny':
            outputs = tinyo
            
        elif data_name == 'vgg':
            outputs = vggo

        test_loss += loss.data[0]
        _, predicted = torch.max(outputs.data, 1)
        total += targets.size(0)
        correct += predicted.eq(targets.data).cpu().sum()
        incorrect += predicted.ne(targets.data).cpu().sum()
        prec1, prec5 = accuracy(outputs.data, target, topk=(1, 5))
        top1.update(prec1[0], inputs.size(0))
        top5.update(prec5[0], inputs.size(0))

    nTotal = len(valloader[data_name].dataset)
    err  = 100.*incorrect/nTotal
    test_loss /= len(valloader[data_name])
    
    print('\nTest set: Average loss: {:.4f}, Error: {}/{} ({:.0f}%)\n'.format(
        test_loss, incorrect, nTotal, err))
    print(' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}'
          .format(top1=top1, top5=top5))
    testF.write('{},{},{},{}\n'. format(epoch, test_loss, top1.avg, top5.avg))
    testF.flush()
    return k[data_name]*err
#    # Save checkpoint.
#    acc = 100.*correct/total
#    
#    state = {
#        'net': net.module if use_cuda else net,
#        #'net': net,
#        'acc': acc,
#        'epoch': epoch,
#    }
#    torch.save(state, os.path.join(args.save, 'cifar.t7'))
#    if acc > best_acc:
#        torch.save(state, os.path.join(args.save, 'cifar.t7'))
#        best_acc = acc 
            

######################################################################
# Visualizing the model predictions
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#
# Generic function to display predictions for a few images
#

def accuracy(output, target, topk=(1,)):
    """Computes the precision@k for the specified values of k"""
    maxk = max(topk)
    batch_size = target.size(0)

    _, pred = output.topk(maxk, 1, True, True)
    pred = pred.t()
    correct = pred.eq(target.view(1, -1).expand_as(pred))

    res = []
    for k in topk:
        correct_k = correct[:k].view(-1).float().sum(0)
        res.append(correct_k.mul_(100.0 / batch_size))
    return res

######################################################################
# Finetuning the convnet
# ----------------------
# Model
if args.resume:
    # Load checkpoint.
    print('==> Resuming from checkpoint..')
    #assert os.path.isdir('checkpoint'), 'Error: no checkpoint directory found!'
    checkpoint = torch.load(os.path.join(args.save, 'cifar.t7'))
    model_ft = checkpoint['net']
    best_acc = checkpoint['acc']
    start_epoch = checkpoint['epoch']
else:
    print('==> Building new model..')
    model_ft = Resnet_sen_share(PreActBlock, [2,2,2,2], 100, 1000, 102)
    #net = ResNet34()
    print ('Number of parames:{}'.format(
        sum([p.data.nelement() for p in model_ft.parameters()])))
    
if use_gpu:
    model_ft = model_ft.cuda()
    model_ft = torch.nn.DataParallel(model_ft, device_ids=[0])
    cudnn.benchmark = True

criterion = nn.CrossEntropyLoss()

######################################################################
# Train and evaluate
# ^^^^^^^^^^^^^^^^^^

if args.resume:
    trainF = {x: open(os.path.join(args.save, 'train_' + x + '.csv'), 'a')
    for x in ['cifar', 'tiny', 'vgg']
    }
    testF = {x: open(os.path.join(args.save, 'test_' + x + '.csv'), 'a')
    for x in ['cifar', 'tiny', 'vgg']
    }
else:
    trainF = {x: open(os.path.join(args.save, 'train_' + x + '.csv'), 'w')
    for x in ['cifar', 'tiny', 'vgg']
    }
    testF = {x: open(os.path.join(args.save, 'test_' + x + '.csv'), 'w')
    for x in ['cifar', 'tiny', 'vgg']
    }  
    
elapsed_time = 0
#epochs = 300
for epoch in range(start_epoch, start_epoch+epochs):
    err = 0
    for data_name in ['cifar', 'tiny', 'vgg']:    
        start_time = time.time()
        train(model_ft, criterion, epoch, trainF[data_name], data_name)
        err += test(model_ft, epoch, testF[data_name], data_name)
        epoch_time = time.time() - start_time
        elapsed_time += epoch_time
    print('| Elapsed time : %d:%02d:%02d'  %(cf.get_hms(elapsed_time)))
    #    # Save checkpoint.
#    acc = 100.*correct/total
#    
    state = {
        'net': model_ft.module if use_cuda else model_ft,
        #'net': net,
        'acc': 100-err,
        'epoch': epoch,
    }
    print (100-err)
    print (best_acc)
#    torch.save(state, os.path.join(args.save, 'cifar.t7'))
    if (100-err) > best_acc:
        torch.save(state, os.path.join(args.save, 'cifar.t7'))
        best_acc = 100-err 

for x in ['cifar', 'tiny', 'vgg']:  
    trainF[x].close()
    testF[x].close()

**Comment by bob48523** [comment]
import random

......

for epoch in range(epochs)
   train()

......

train ():
   N1 = size(cifar)
   N2 = size(tiny)
   N3 = size(vgg)
   N = (N1+N2+N3)/batch_size
   batch_indx = [i for i in range(N)]
   random.shuffle(batch_indx)

   for indx in batch_indx:
      if indx >= N1+N2:
         indx = indx-N1-N2
         data_name = 'vgg'
      elif indx >= N1:
         indx = indx-N1
         data_name = 'tiny'
      else:
         data_name = 'cifar'

      inputs, targets = trainloader[data_name][indx]
      ......
